Lesson5a

paulcraig opened the pull request:
This is the linear regression lesson. I would like to learn how to format tables better in Jupyter book.

paulcraig committed:
corrected a few typos

paulcraig committed:
first half of the file

paulcraig committed:
adding PDB files to the site

paulcraig committed:
small molecule file added

paulcraig committed:
full draft of lesson 2

paulcraig committed:
sample output file for lesson 3

paulcraig committed:
full draft of lesson 3

paulcraig committed:
sample written file from lesson 3

paulcraig committed:
sample resolutions.txt output file for lesson 3

paulcraig committed:
adding code for linear regression

paulcraig committed:
corrected some typos

paulcraig committed:
included scipy and seaborn linear regression

paulcraig committed:
data for the Bradford protein assay

paulcraig committed:
changes in formatting to fit in the Jupyter book

paulcraig committed:
rewrote the pandas exercise with a protein csv file

paulcraig committed:
adding more data files for the pandas exercise

paulcraig committed:
changed the format for the exercises

paulcraig committed:
output for the final exercise

paulcraig committed:
updated the table of contents

paulcraig committed:
changing the format for Jupyter book

paulcraig committed:
added File Parsing to the TOC

paulcraig committed:
corrected one typo

paulcraig committed:
corrected typos and formatting errors

paulcraig committed:
full draft with output cleared

paulcraig committed:
split the notebook in 3 parts - this is lin reg

paulcraig committed:
split the notebook in 3 parts - thisis pandas

paulcraig committed:
split the notebook in 3 parts - thisis plotting

paulcraig committed:
first draft

paulcraig committed:
removed the help(pd.read_csv) cell

paulcraig committed:
changed the help(read_csv) output

paulcraig committed:
full draft of the pandas lesson

paulcraig committed:
data file for the exercise

paulcraig committed:
image for Working with Pandas

paulcraig committed:
added an exercise with ground water pH

paulcraig committed:
fixed formatting errors

paulcraig committed:
fixed formatting errors

paulcraig committed:
data file for LR exercise

paulcraig committed:
corrected formatting and typos

paulcraig committed:
changed sub-headings

paulcraig committed:
run and clear output

janash commented:
How would you like the tables to be formatted? You have used markdown formatting for tables, which is perfectly fine.

However, since jupyter books and jupyter notebooks use HTML and CSS, you can format tables any way you would like as long as you have the CSS styles for it. What kind of formatting were you wanting?

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,230 @@
+    "# Import the libraries you need\n",
+    "import os\n",
+    "import pandas as pd\n",
+    "import scipy as sp"
```
Need more of an explanation of scipy.

paulcraig replied:
I expanded the explanation of SciPy based on a nice source on the web.

paulcraig replied:
I can't see where I can respond to your comments on table formatting, so I'll add that comment here. In the Jupyter Book they appear to be double spaced. I would like them to be single spaced and shaded a bit, the way they appear in the Jupyter notebook. It's not urgent, but it would be nice.

paulcraig committed:
adding the two PDB files for the workbook

paulcraig committed:
with suggested corrections

paulcraig committed:
Delete File_Parsing.ipynb

paulcraig committed:
Delete introduction.ipynb

paulcraig committed:
Delete remdesivir.pdb

paulcraig committed:
Delete remdesivir.sdf

paulcraig committed:
Delete remdesivir.xyz

paulcraig committed:
Delete 1vxh.pdb

paulcraig committed:
Delete 4eyr.pdb

paulcraig committed:
Delete 4ezi.pdb

paulcraig committed:
Delete 6l35.pdb

paulcraig committed:
Delete 6lsv.pdb

paulcraig committed:
Delete 6vqs.pdb

paulcraig committed:
Delete 7bfv.pdb

paulcraig committed:
Delete 7bgj.pdb

paulcraig committed:
Delete 7dno.pdb

paulcraig committed:
Delete 7l1f.pdb

paulcraig committed:
Delete 7tim.pdb

paulcraig committed:
Delete remdesivir.xyz

paulcraig committed:
made all changes; ran and cleared output

paulcraig committed:
adding all enzymes to the PDB_files folder

paulcraig committed:
Delete 2pkr.pdb

paulcraig committed:
Delete 3iva.pdb

paulcraig committed:
Delete 3vnd.pdb

paulcraig committed:
Delete 5eu9.pdb

paulcraig committed:
Delete 5veu.pdb

paulcraig committed:
Delete 6zt7.pdb

paulcraig committed:
Delete 7tim.pdb

paulcraig committed:
updated TOC

paulcraig committed:
changes made, restarted and cleared output

paulcraig committed:
Delete Processing_Multiple_Files_and_Writing_Files.ipynb

Something wrong with my git operations. I am deleting this file on Github and will try to send it from my local repository.

paulcraig committed:
written file from the main notebook

paulcraig committed:
cleaning up unused files from lesson4a

paulcraig committed:
added a new check your understanding

paulcraig committed:
added a file with multiple HETNAM entries

paulcraig committed:
current state of lesson 3

paulcraig committed:
copy of lesson 3

paulcraig committed:
removing non-linear regression files

paulcraig committed:
removing more extraneous files

paulcraig committed:
removed introduction.ipynb from lesson5a branch

paulcraig committed:
added required PDB files

paulcraig committed:
cleaning out unnecessary files

paulcraig committed:
updated TOC

paulcraig committed:
added information about SciPy

paulcraig committed:
protein unknown calculations

paulcraig committed:
restoring the protein data file

paulcraig committed:
protein unknowns added

paulcraig committed:
adding protein unknown calculations

paulcraig committed:
refining the protein sample analysis

paulcraig committed:
not ready for prime time draft

paulcraig committed:
full draft, ran and cleared output

paulcraig commented:
I added some markdown and code cells for determining the concentrations of unknowns. I did some things with pandas that I had to learn about online. Please let me know what you think.

1. Are they too advanced?
2. Are they ugly coding that we don't want to teach others to do?
3. In the section entitled, Eliminating values outside the calibration curve (optional), I used two conditionals to make changes in the dataframe. I wanted to use both conditionals in one line, but could not figure out how to do it without generating errors.

paulcraig committed:
minor text changes

janash committed:
Merge branch 'lesson3a' of https://github.com/paulcraig/python-scripting-biochemistry into paulcraig-lesson3a

janash committed:
slight clean up of notebook, delete unneccessary file

janash committed:
Merge pull request #4 from paulcraig/lesson3a

Lesson3a

janash committed:
Merge branch 'main' into lesson2a

janash committed:
Merge pull request #3 from paulcraig/lesson2a

Lesson2a

janash committed:
Merge branch 'lesson4a' of https://github.com/paulcraig/python-scripting-biochemistry into paulcraig-lesson4a

janash committed:
fix merge conflicts on branch

paulcraig committed:
made all of the requested changes

paulcraig committed:
added the required csv file

paulcraig committed:
added Working_with_Pandas.ipynb to the TOC

paulcraig committed:
minor adjustments for Jupyter book

paulcraig committed:
adding data files

janash committed:
minor formatting changes

janash committed:
Merge pull request #5 from paulcraig/lesson4a

Lesson4a

janash committed:
fix overview div at top of pandas lesson

janash committed:
switch image to directive

janash committed:
Merge pull request #10 from janash/overview-fix

Overview fix

janash committed:
fix merge conflicts with main branch

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,359 @@
+   "source": [
+    "from scipy import stats\n",
+    "slope, intercept, r_value, p_value, std_err = stats.linregress(xdata, ydata)\n",
+    "print(\"Slope = \", slope, \"/microgram/mL\", sep = \"\")\n",
```
Where did `mL` come from here? This unit is what you would get if you were fitting absorbance vs. concentration, but as described in the previous cells, the y-value is mass (meaning the unit of slope would be `mass^-1` to make the y-value unit-less). I'm more accustomed to calibration curves being done with concentrations, so maybe you meant that the units should be `ug/mL` in the dataframe rather than just `ug`?

I'm assuming this calibration curve would be done with a standard volume (all samples the same), so the concentration and the mass is directly proportional.

paulcraig replied:
Thanks for pointing this out. I have been using my own, somewhat quirky approach to protein calibration curves for years. I checked the literature and the manuals and they all use protein concentration, rather than protein mass. I changed the units and the markdown to reflect protein concentration for the standards and the samples.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,359 @@
+    "Of course, we have the equation for our line\n",
+    "\n",
+    "$$\n",
+    "A_{595} = slope * ProteinMass + intercept\n",
```
If the slope is in `mL/microgram`, this should be protein concentration instead of mass. We would use it along with the sample volume to solve for 

A_{595} = slope * ProteinConcentration + intercept

if the units for slope are mL/microgram, units for x should be microgram/mL so that A_{595} is unit-less.

paulcraig replied:
With the x axis as mg/mL, the slope units become mL/mg. The units for protein concentration are mg/mL. So A\_{595} is unitless as it should be.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,359 @@
+    "\n",
+    "| Sample volume | A_595 |\n",
+    "| :------------: | :----: |\n",
+    "| ($\\mu$L) | |\n",
```
Suggest moving the unit to be in the same cell as the header.

paulcraig replied:
When I changed the units to mg/mL for concentration, I also fixed the sample volume for standards and samples at 100 microliters. So now the only columns in the csv as sample # and A-595. No need for units.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,359 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "samples_df.loc[samples_df['A-595'] < 0.094425, 'Concentration'] = 'Out of Range'\n",
```
If you use the string `Out of Range` here, it will change the data type of the column from float to string. To preserve the column staying a float, you could use `np.nan` (not a number). This is typically what is used when there is no value for the cell.

paulcraig replied:
Thanks. I never would have thought of that and now I understand why I saw so many NaN values in Corey Schafer's videos about pandas.

janash reviewed (changes requested):
The data in this lesson is confusing to me. Please address the following questions. Minor coding comments as well.

I am confused about the units for the linear fit. This fits absorbance vs mass, which is strange to do to to me. I've more often seen absorbance vs concentration (although mass and concentration would typically be directly proportional because standard cuvette volumes are used for absorbance measurements).  The slope units are given as /micrograms/mL, and if the x values is mass it should only be /micrograms. 

The current calculations don't make sense to me. It seems like the original fit should be done with x being the protein concentration rather than the mass. 

Before changing any of the files on this lesson, make sure to pull the changes I've made to your local branch.

```
git checkout lesson5a
git pull origin lesson5a
```

paulcraig committed:
data file for the exercise

paulcraig committed:
latest draft of the notebook plus the csv files

paulcraig commented:
I believe I addressed all of your suggestions and concerns. It all appears to work as a local Jupyter book build.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,752 @@
+    "import os\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import scipy as sp"
```
I noticed that we don't actually use this import in the notebook. Later we do `from scipy import stats`, but we don't actually use `sp` at all. I suggest changing the import at the start of the notebook to `from scipy import stats`, or not importing scipy until you are ready to do the fit.

paulcraig replied:
I added from scipy import stats on line 89. Then I deleted from scipy import stats from the cell where the regression is first called.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,752 @@
+    }
+   ],
+   "source": [
+    "samples_df['ProtConc'] = ((samples_df['A-595'] - intercept) / slope)\n",
```
You could also set it equal to the list `ProtConc` you created in the previous cell.

paulcraig replied:
I made the change and I like it. It's cleaner.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,752 @@
+    }
+   ],
+   "source": [
+    "ProtConc = ((samples_df['A-595'] - intercept) / slope)\n",
```
Recommend different variable name here (`protein_concentration` or `protein_conc`). This variable naming style (all lowercase with underscores separating words) is conventional in Python.

paulcraig replied:
I used protein_conc.

janash commented on biochemist-python/chapters/Linear_Regression.ipynb:
```diff
@@ -0,0 +1,752 @@
+    "samples_df.loc[samples_df['A-595'] < 0.642, 'ProtConc'] = np.nan\n",
+    "samples_df.loc[samples_df['A-595'] > 1.188, 'ProtConc'] = np.nan\n",
+    "samples_df"
+   ]
```
After showing this, it could also be an opportunity to introduce how to do this programmatically instead of using the hard-coded numbers of `0.642` and `1.188`. If you didn't want to look these up, you could use the `min` and `max` function on the data you used to fit the linear equation, ie

```
samples_df.loc[samples_df['A-595'] < results_df['A-595'].min(), 'ProtConc'] = np.nan
```

paulcraig replied:
I added a cell after this one and just before the exercise where I mentioned the .min() and .max() functions.

janash reviewed (changes requested):
Mostly minor comments to look over.

paulcraig commented:
I incorporated your suggestions. I also deleted the last cell that contained the solution to the exercise. I did a Jupyter book build and it looks okay.

paulcraig committed:
made the changes requested in the pull request

janash committed:
minor formatting change
